cudamapper/src/utils.cpp

```diff
@@ -150,6 +156,182 @@ std::vector<IndexDescriptor> group_reads_into_indices(const io::FastaParser& par
+
+        // write RNEXT, PNEXT, and TLEN
+        // all left as blank/default for now
+
+        // write SEQ
+        encode_seq(alignment, query_parser.get_sequence_by_id(overlaps[i].query_read_id_).seq);
+
+        // QUAL is set to * by default
+
+        // TODO: write AUX data if available
+
+        std::lock_guard<std::mutex> lg(write_output_mutex);
+        int out = sam_write1(file, header, alignment);
+        bam_destroy1(alignment);
+    }
+
+    // destroy header, and close fd to stdout
+    sam_hdr_destroy(header);
+    sam_close(file);
+    return;
+}
+
+static inline void encode_cigar(bam1_t* alignment, const std::string& cigar)
+{
+    uint32_t* bam_cigar_start = bam_get_cigar(alignment);
+    memcpy(bam_cigar_start, cigar.c_str(), cigar.length());
+    alignment->core.n_cigar = cigar.length();
+
+    return;
+}
+
+static inline void encode_seq(bam1_t* alignment, const std::string& seq)
+{
+    // This table was taken from bam_construct_seq() for base -> nibble encoding
+    static const char L[256] =
+        {
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 0, 15, 15,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15};
+
+    const size_t len = seq.length();
+
+    // encode bases as nibbles and write them into the alignment struct 2 bases at a time
+    // 4 high bits are first base, 4 low bits are second base
+    uint8_t* bam_seq_start = bam_get_seq(alignment);
+    int i                  = 0;
+    for (i; (i + 1) < len; i += 2)
+    {
+        *bam_seq_start++ = (L[(unsigned char)seq[i]] << 4) + L[(unsigned char)seq[i + 1]];
+    }
+    // odd number of bases, so we have one left over
+    if (i < len)
+    {
+        *bam_seq_start++ = L[(unsigned char)seq[i]] << 4;
+    }
+
+    alignment->core.l_qseq = len;
+
+    *bam_seq_start = 0xff; // if htslib encounters this at the start of the QUAL section, it writes default value (*)
```

Comment:
let's move this into an `encode_quality` function just to be explicit about it

Reply:
done